tmol/score/omega/potentials/omega_pose_score.impl.hh

```diff
@@ -0,0 +1,206 @@
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+
+#include <tmol/utility/tensor/TensorAccessor.h>
+#include <tmol/utility/tensor/TensorPack.h>
+#include <tmol/utility/tensor/TensorStruct.h>
+#include <tmol/utility/tensor/TensorUtil.h>
+#include <tmol/utility/nvtx.hh>
+
+#include <tmol/score/common/accumulate.hh>
+#include <tmol/score/common/count_pair.hh>
+#include <tmol/score/common/data_loading.hh>
+#include <tmol/score/common/diamond_macros.hh>
+#include <tmol/score/common/geom.hh>
+#include <tmol/score/common/launch_box_macros.hh>
+#include <tmol/score/common/sphere_overlap.impl.hh>
+#include <tmol/score/common/tile_atom_pair_evaluation.hh>
+#include <tmol/score/common/tuple.hh>
+#include <tmol/score/common/warp_segreduce.hh>
+#include <tmol/score/common/warp_stride_reduce.hh>
+
+#include <tmol/score/omega/potentials/omega_pose_score.hh>
+
+// Operator definitions; safe for CPU compilation
+#include <moderngpu/operators.hxx>
+
+#include <chrono>
+
+#include "potentials.hh"
+
+// The maximum number of inter-residue chemical bonds
+#define MAX_N_CONN 4
+#define TILE_SIZE 32
+
+namespace tmol {
+namespace score {
+namespace omega {
+namespace potentials {
+
+template <typename Real, int N>
+using Vec = Eigen::Matrix<Real, N, 1>;
+template <typename Real>
+using CoordQuad = Eigen::Matrix<Real, 4, 3>;
+
+template <typename Real, typename Int, tmol::Device D>
+TMOL_DEVICE_FUNC int resolve_atom_from_uaid(
+    TensorAccessor<Int, 1, D> uaid,
+    int block_index,
+    int pose_index,
+
+    TView<Vec<Real, 3>, 2, D> coords,
+    TView<Int, 2, D> pose_stack_block_coord_offset,
+    TView<Int, 2, D> pose_stack_block_type,
+    TView<Vec<Int, 2>, 3, D> pose_stack_inter_block_connections,
+    TView<Int, 3, D> block_type_omega_quad_uaids,
+    TView<Int, 3, D> block_type_atom_downstream_of_conn) {
+  if (uaid[0] != -1) {  // This uaid resides in this block
+    int block_coord_offset =
+        pose_stack_block_coord_offset[pose_index][block_index];
+    return block_coord_offset + uaid[0];  // The actual global atom index
+  } else {                                // We need to follow to another block
+    int connection_index = uaid[1];
+    int sep = uaid[2];
+
+    const Vec<Int, 2>& connection =
+        pose_stack_inter_block_connections[pose_index][block_index]
+                                          [connection_index];
+    int other_block_index = connection[0];
+    int other_connection_index = connection[1];
+    int other_block_type_index =
+        pose_stack_block_type[pose_index][other_block_index];
+
+    int idx = block_type_atom_downstream_of_conn[other_block_type_index]
+                                                [other_connection_index]
+                                                [sep];  // The offset within the
+                                                        // other block
+    int block_coord_offset =
+        pose_stack_block_coord_offset[pose_index][other_block_index];
+    return block_coord_offset + idx;
```

Comment:
@aleaverfay This is where I suspect things might be producing non-deterministic behavior since I do nothing to check if a connection is valid.